### PR TITLE
Build with Meson

### LINKED ISSUE #79: Port to Meson
Would benefit greatly from conditionals like mapping default apps to the different names of apps across distros. Tricky part is how to feed `dh_installgsettings` the overrides once they're moved out of the `debian/` dir.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58527170-port-to-meson?utm_campaign=plugin&utm_content=tracker%2F74974394&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F74974394&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by worldofpeace**
Fixes #79 

Should essentially complete https://github.com/elementary/default-settings/tree/meson.

### CODE REVIEW

**Comment by worldofpeace** [comment]
I've rebased to master, and added a commit to install the accountsservice files https://github.com/elementary/default-settings/pull/119/commits/89e53b6090b8981cab5f4d97cd3c5759b10c47fd.

**Comment by worldofpeace** [comment]
@codygarver I'd really like to get this merged. I've rebased this for nth time..

**Comment by quequotion** [comment]
TravisCI fails with:
```
Called `/usr/bin/pkg-config --modversion polkit-gobject-1` -> 1
1604
1605Dependency polkit-gobject-1 found: NO (tried pkgconfig)
1606
1607accountsservice/meson.build:10:0: ERROR:  Dependency "polkit-gobject-1" not found, tried pkgconfig
```

In Archlinux this is provided by the `polkit` package; not sure about Ubuntu/Debian packaging (`polkit-gobject`?). Somehow need to get TravisCI to preinstall the dependency before building the package. I don't remember how this is done, but I think it needs to be done in `.travis.yml`.

**Comment by worldofpeace** [comment]
Last time an elementary OS member reviewed my meson PR's, I was instructed to leave the CI stuff to them. And I think they're switching from Travis CI anyway.

**Comment by btd1337** [comment]
Any progress on this?

**Comment by btd1337** [comment]
Travis CI is pointed to unstable image. Is this okay?

`- docker run -v "$PWD":/tmp/build-dir elementary/docker:juno-unstable`

**Comment by worldofpeace** [comment]
> Any progress on this?

I can rebase this. It's been complete for a while now, and I've use it downstream in NixOS.

**Comment by worldofpeace** [comment]
> Travis CI is pointed to unstable image. Is this okay?
>- docker run -v "$PWD":/tmp/build-dir elementary/docker:juno-unstable

They use GitHub for CI now. I've switched it over in this PR.

**Comment by worldofpeace** [comment]
Perhaps @decathorpe could review this.

**Comment by decathorpe** [comment]
@worldofpeace Sorry, I seem to have missed the notification to review this PR.

I'm probably not the best person to do it though, since I'm not using this project at all.

**Comment by worldofpeace** [comment]
@danrabbit Conflicts resolved.

**Comment by worldofpeace** [comment]
CI failure here is anyone's guess
```
 Dependency accountsservice found: YES 0.6.45
WARNING: pkgconfig variable 'interfacesdir' not defined for dependency accountsservice.
Program create-symlink.py found: YES (/usr/bin/env python3 /__w/default-settings/default-settings/accountsservice/create-symlink.py)
Build targets in project: 0
Found ninja-1.8.2 at /usr/bin/ninja
ninja: Entering directory `build'
ninja: no work to do.
ninja: Entering directory `build'
[0/1] Installing files.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mesonbuild/mesonmain.py", line 112, in run
    return options.run_func(options)
  File "/usr/lib/python3/dist-packages/mesonbuild/minstall.py", line 504, in run
    installer.do_install(datafilename)
  File "/usr/lib/python3/dist-packages/mesonbuild/minstall.py", line 340, in do_install
    self.install_subdirs(d) # Must be first, because it needs to delete the old subtree.
  File "/usr/lib/python3/dist-packages/mesonbuild/minstall.py", line 361, in install_subdirs
    self.do_copydir(d, src_dir, full_dst_dir, exclude, mode)
  File "/usr/lib/python3/dist-packages/mesonbuild/minstall.py", line 288, in do_copydir
    raise ValueError('dst_dir must be absolute, got %s' % dst_dir)
ValueError: dst_dir must be absolute, got out/etc/skel/.config/plank
Installing subdir /__w/default-settings/default-settings/plank to out/etc/skel/.config/plank
FAILED: meson-install 
/usr/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
```

**Comment by worldofpeace** [comment]
cc @jtojnar if you're interested.
I'm planning to do some sort of followup to this where the plank dockitems have correct exec's on whatever platform.

**Comment by decathorpe** [comment]
@worldofpeace if you don't set DESTDIR, I think meson will try to install to /usr/local by default, which should fail :/ That's why I included the DESTDIR environment variable for all CI projects when I first included the "install" step in the test run

I have no idea why a relative path doesn't work in this case - but setting it to `$(pwd)/build/out` instead of `out` should make DESTDIR absolute instead of relative, while keeping the effect the same? No idea whether GitHub Actions support that, though.

**Comment by jtojnar** [comment]
https://github.com/mesonbuild/meson/issues/5207

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,79 @@
+project(
+    'elementary-default-settings',
+    version: '5.1.0',
+    default_options: 'sysconfdir=/etc',
```
Comment (jtojnar) [inline meson.build:4]: Why this? It should be the default value if you also use the default prefix.

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+prefix = get_option('prefix')
+datadir = join_paths(prefix, get_option('datadir'))
+sysconfdir = get_option('sysconfdir')
+
+meson.add_install_script('meson/post_install.py', datadir)
```
Comment (jtojnar) [inline meson.build:11]: Maybe pass schemadir to allow reuse.

```
schemadir = datadir / 'glib-2.0' / 'schemas'
```

File: overrides/meson.build [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    default_wallpaper = get_option('default-wallpaper')
+
+    wallpaper_configuration = configuration_data()
+    wallpaper_configuration.set('DEFAULT_WALLPAPER', default_wallpaper)
+
+    settings_override = configure_file(
+        input: 'default-settings.gschema.override.in',
+        output: '@BASENAME@',
+        configuration: wallpaper_configuration,
+        install_dir: join_paths(datadir, 'glib-2.0', 'schemas')
```
Comment (jtojnar) [inline overrides/meson.build:11]: schemadir

File: meson.build [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+# Default Appcenter Blacklist
+install_data(
+    'appcenter/appcenter.blacklist',
+    install_dir: join_paths(sysconfdir, 'appcenter')
+)
+
+# Configuration file of GNU readline
+install_data(
+    'inputrc/.inputrc',
+    install_dir: join_paths(sysconfdir, 'skel')
```
Comment (jtojnar) [inline meson.build:72]: Maybe group the skeletor things together?

**Comment by worldofpeace** [comment]
> @worldofpeace if you don't set DESTDIR, I think meson will try to install to /usr/local by default, which should fail :/ That's why I included the DESTDIR environment variable for all CI projects when I first included the "install" step in the test run
> 
> I have no idea why a relative path doesn't work in this case - but setting it to `$(pwd)/build/out` instead of `out` should make DESTDIR absolute instead of relative, while keeping the effect the same? No idea whether GitHub Actions support that, though.

I see, I will keep that in mind when adding CI to projects here that don't have it already :+1: 
I will try doing this in another push.

**Comment by worldofpeace** [comment]
Travis CI will always be expectant here, since it's used on master.
CI does pass in the github action.

**Comment by danrabbit** [comment]
@worldofpeace I'd like to move this forward, but it requires packaging changes. Can you review/merge this branch? https://github.com/worldofpeace/default-settings/pull/1

**Comment by worldofpeace** [comment]
> @worldofpeace I'd like to move this forward, but it requires packaging changes. Can you review/merge this branch? [worldofpeace#1](https://github.com/worldofpeace/default-settings/pull/1)

Woah, it looks like you opened that a long time ago. Sorry about that. My notifications are out of control :smile: I can try to integrate this.

**Comment by worldofpeace** [comment]
How can I test the packaging for debian?

**Comment by danrabbit** [comment]
Testing locally with `debuild -i -us -uc -b`, it looks like we still have to fix some things in packaging:

- [ ] the symlink in `/usr/share/accountsservice/interfaces/` is missing
- [ ] `default-settings.gschema.override` in `usr/share/glib-2.0/schemas/` is missing

**Comment by danrabbit** [comment]
I'm getting a warning: `WARNING: pkgconfig variable 'interfacesdir' not defined for dependency accountsservice`

**Comment by worldofpeace** [comment]
> I'm getting a warning: `WARNING: pkgconfig variable 'interfacesdir' not defined for dependency accountsservice`
> 
> Looks like the other files probably aren't created because meson isn't building anything upside_down_face

I grabbed a .deb from ubuntu with pkgconfig file and it appears to have that entry.

**Comment by danrabbit** [comment]
@worldofpeace which version of Ubuntu? The deb I grabbed for bionic seems to not have it

**Pull request merged**